### ISSUE TITLE
Cron run early when changing CDS to an hour past that of the last cron

### ISSUE THREAD

**Issue by paglias** [HabitRPG/habitica#4488]
To reproduce:
1. User has `dayStart` set to `0` and logins at `1am` of `Jan 1`
2. On `Jan 1` at `2am` user changes `dayStart` to `4am`
3. When the user perform an action on `Jan 1` past `4am` the system will see a `lastCron` that happened before `4am`, it will think of it as yesterday and run cron.

**Possible solution:**

Store the old `dayStart` value and use it to calculate if cron should run here https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/index.coffee#L72

**Comment by brandonjreid** [comment]
I can take a look at this later this week.

**Comment by brandonjreid** [comment]
Sorry, I got busy and wasn't able to look at this for a while.

I'm wondering if a better solution might be that when dayStart is set to a new value, check lastCron vs old_dayStart.  If old_dayStart < lastCron < new_dayStart, then set lastCron to new_dayStart.  That way cron won't run twice for one day (1 am and after 4 am based on your example).  It seems like a simpler solution than storing an extra variable that we have to check every time daysSince is called (about once per request).  However, I don't know if it would be difficult to get a hold of the lastCron variable in that part of the code.  I'm not too familiar with that area.

**Comment by Alys** [comment]
Can anyone think of any problems with @brandonjreid's suggestion? It seems reasonable to me. If there's no comments in another couple of days, I'd suggest that @brandonjreid goes ahead with it, if he'd like to.

**Comment by crookedneighbor** [comment]
Sounds reasonable to me. @brandonjreid, both of those variables are stored on the user object, so it should be very easy to manipulate both.

**Comment by Alys** [comment]
@brandonjreid Would you be interested in implementing your suggestion? I think it would be a great improvement.

**Comment by carolstone** [comment]
Is anyone working on this? I'm very willing to try, if nobody is knee-deep in it yet.

**Comment by brandonjreid** [comment]
Go for it!  I've been meaning to get around to it, but I've been busy, so I haven't had a chance to look at it.  Feel free to work on it.

**Comment by carolstone** [comment]
Thank you. I'll probably have a bunch of questions, but I would really like to get this fixed.

**Comment by carolstone** [comment]
The logic to be added is: If old_dayStart < lastCron < new_dayStart, then set lastCron to new_dayStart 

It seems to me this logic belongs here: https://github.com/HabitRPG/habitrpg/blob/develop/website/public/js/controllers/settingsCtrl.js#L69-L76

I can get lastCron by setting a variable to User.user.lastCron, and updating lastCron appears straightforward enough.

What I can't seem to figure out is how to get the previous dayStart value. It's bound to an ng-model here: https://github.com/HabitRPG/habitrpg/blob/develop/website/views/options/settings.jade#L86 so when I look at the value of User.user.preferences.dayStart I only get the value that has been changed but not written to the database.

All the advice I can find about how to get the value of dayStart before the value in the input was touched leads me to request someone with a clear understanding of the design decisions on this project to give me guidance on the preferred way to proceed. Is it acceptable to add a $watch to this code? $watch is supposed to be bad for performance. All the non-$watch solutions I've found appear very cumbersome and considerably less than ideal in this situation. I think we got to this solution in a sensible attempt to avoid creating another variable in the user object, so I'm not of a mind to go there, either. Has this type of problem been solved elsewhere on the site with a strategy I should copy?

**Comment by crookedneighbor** [comment]
I would add a variable on the settings controller $scope that initialized with the value of the user's dayStart. Put that on the ng-model instead. Then in the save day start function, you can compare the two.

**Comment by carolstone** [comment]
@crookedneighbor  That makes a lot more sense than anything I was looking at. Thank you. I'll pick up there tomorrow (well, later today).

**Comment by carolstone** [comment]
I just want to check in and let anyone who's interested know that I'm still working on this. I'm actually getting pretty far along, but I'm brand-new at all of the technologies used here, so it's taking me a long time to do things that would be trivial for most people here.

**Comment by crookedneighbor** [comment]
@carolstone Awesome. If you get stuck, feel free to open a PR as a work in progress and we'll help you out.

**Comment by carolstone** [comment]
Well, I've issued a pull request. I was much too verbose explaining what I did. And I referenced this issue. Please accept my apologies - I think I was supposed to rebase before issuing the pull request, but this is so new to me, I stopped where I knew it was working. Please scrutinize my work. This is my first pull request ever. I've never written any jade or coffeescript before, and I've only done a few practice things in javascript. I am, however, a seasoned developer (18 years), so I'm not completely lost!

**Comment by carolstone** [comment]
I've got it working and I really don't like this solution. I see two problems with it. First of all, it breaks the [8-days a week challenge](https://habitrpg.com/#/options/groups/challenges/7fffa2f4-6e45-4232-b061-f814c448fa17) by making one day last up to 47 hours. Second, cron will run an hour apart if you set your day offset from 23 to 0.

I thought I should make up a spreadsheet proposing a different solution, as well as a spreadsheet to show exactly how I think this will break the 8 days a week challenge.

My proposed different solution is to restrict the amount of time that passes between lastcron and the newly calculated daystart, the idea being to allow someone to move cron around without allowing someone to force cron to run within whatever we define as too small a number of hours, either by accident or on purpose. (I know it will always be possible for someone to log in at 11:15 pm, run cron, and then have it run again at midnight. I'm not proposing to fix that. I'm just saying we can prevent people from doing it with the day offset.)

My life is rather nuts right now, since I have several major deadlines on July 31 and one on August 3 (and another three on August 15. But I'll give this some time anyhow and see if I can't get at least some kind of example posted for consideration sooner rather than later.

**Comment by Alys** [comment]
@carolstone In your pull request, could you please edit the title and enter "Fixes #4488"? That will create an automatic link between this issue and the pull request.

I'm not too concerned about breaking that challenge. The challenge is used by only 30 people, and this cron problem affects many more users than that.

We would definitely like to see details of your proposal, but keep in mind that there are legitimate reasons for cron to run twice in a short space of time, regardless of what CDS is set (e.g, as you say, if the user doesn't log in on one day until shortly before their next day starts), so we'll need to be cautious about preventing that from happening.

**Comment by carolstone** [comment]
I thought I had this working. I _do_ have this working. But I set up a test with a quest. It turns out, if your daystart is set for 22 (10:00 p.m.) and you log in at 10:30 pm, cron runs. If you then update daystart to be 23 (11:00 p.m.) and push last cron to be 11:00 p.m. cron runs again and you get a message in your quest about it. I made sure it was setting it to 11:00 p.m. in the future, not in the past (at first I thought it was setting it to yesterday, but it's not). Sometimes I had to select sync to get it to run, but it always runs. I think I need to study how cron runs a little more closely. Sorry.

**Comment by Alys** [comment]
@carolstone Would you like us to have a look at the code, if you've made any changes since your last push? If so, you can push the modified code to your chron-run-early-issue-4488 branch and it will appear in your PR.

**Comment by carolstone** [comment]
Ok. I'll check in my code. I'll leave my alerts in for the moment. I'm posting below a picture of tonight's test run. I used "cron" and "daystart" interchangeably because when I reset the daystart it was resetting cron. The only thing I left out was changing daystart to 20 shortly before running sync at 9:21 p.m. to show that nothing happened. The notes of what I did come after I did the thing. 
![capturecron](https://cloud.githubusercontent.com/assets/9287541/9102584/1923a016-3bc2-11e5-9c42-36a44c3995aa.JPG)
